MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,140 @@
+Maintenance Job Design Document
+===============================
+
+The (shared) object cache is growing to be large. This includes:
+
+* Number of loose objects.
+* Number of pack-files.
+* Size of pack-files.
+
+There are some things we can do to improve this situation, with varying levels
+of difficulty and benefit. Here are the four items I believe will have a strong
+impact on our data size and performance:
+
+I. Delete loose objects that exist in pack-files.
+II. Place loose objects into pack-files.
+III. Repack some prefetch packs into larger packs.
+IV. Delete packs with no objects referenced by the multi-pack-index.
+
+These are numbered by order of implementation, but they work together in an
+interesting way. If we place loose objects into pack-files (I), then the next
+time we delete loose objects (II) we will delete those objects that we just put
+into a pack-file. If we repack some prefetch packs into larger packs (III), the
+multi-pack-index will use the "newer" pack for all objects in those packs, and
+we can delete them (IV).
+
+In order to perform this maintenance without disrupting users, we can actually
+do these interactions in two independent jobs, with their sub-actions in a
+particular order:
+
+A. Loose Object Maintenance
+  1. Delete loose objects that exist in pack-files.
+  2. Place loose objects into pack-files.

```

Comment:
Maybe this has changed and could be different with the read-object hook but I remember in the git code that it would
1. check pack files
2. check loose object
3. check pack files again

In the comments it said something about objects that may have just been repacked.  Will this keep the loose objects around until we run again and step 1 will delete the loose objects that were moved into pack-files?

Reply:
The comments are regarding missing objects. So Git will follow this mechanism:

1. Check pack-files.
2. check loose object.
3. If first time at this step, check read-object hook. If success, goto 1.
4. If first time at this step, reprepare_packed_git and goto 1. 
5. Fail to find object.

The reprepare_packed_git is intended for Git processes that loaded the packed_git list before a repack finished, but tried to access an object that was previously loose after the repack finishes.

You are correct: we intend to leave loose objects around until the _next_ run of the loose object job. That next run should be far out enough that we are sure no Git processes live across that boundary.